bench-tps/src/bench.rs

```diff
@@ -154,58 +173,40 @@ where
 
     // generate and send transactions for the specified duration
     let start = Instant::now();
+    let keypair_chunks = source_keypair_chunks.len() as u64;
     let mut reclaim_lamports_back_to_source_account = false;
     let mut i = keypair0_balance;
-    let mut blockhash = Hash::default();
-    let mut blockhash_time;
     while start.elapsed() < duration {
-        // ping-pong between source and destination accounts for each loop iteration
-        // this seems to be faster than trying to determine the balance of individual
-        // accounts
-        let len = tx_count as usize;
-        blockhash_time = Instant::now();
-        if let Ok((new_blockhash, _fee_calculator)) = client.get_new_blockhash(&blockhash) {
-            blockhash = new_blockhash;
-        } else {
-            if blockhash_time.elapsed().as_secs() > 30 {
-                panic!("Blockhash is not updating");
-            }
-            sleep(Duration::from_millis(100));
-            continue;
-        }
-        datapoint_debug!(
-            "bench-tps-get_blockhash",
-            ("duration", duration_as_us(&blockhash_time.elapsed()), i64)
-        );
-
-        blockhash_time = Instant::now();
-        let balance = client.get_balance(&id.pubkey()).unwrap_or(0);
-        metrics_submit_lamport_balance(balance);
-        datapoint_debug!(
-            "bench-tps-get_balance",
-            ("duration", duration_as_us(&blockhash_time.elapsed()), i64)
-        );
-
+        let chunk_index = (i % keypair_chunks) as usize;
         generate_txs(
             &shared_txs,
-            &blockhash,
-            &keypairs[..len],
-            &keypairs[len..],
+            &recent_blockhash,
+            &source_keypair_chunks[chunk_index],
+            &dest_keypair_chunks[chunk_index],
             threads,
             reclaim_lamports_back_to_source_account,
             &libra_args,
         );
-        // In sustained mode overlap the transfers with generation
-        // this has higher average performance but lower peak performance
-        // in tested environments.
-        if !sustained {
+
+        // In sustained mode, overlap the transfers with generation. This has higher average
+        // performance but lower peak performance in tested environments.
+        if sustained {
+            // Ensure that we don't generate more transactions than we can handle.
+            while shared_txs.read().unwrap().len() > 2 * threads {
+                sleep(Duration::from_millis(1));
+            }
+        } else {
             while shared_tx_active_thread_count.load(Ordering::Relaxed) > 0 {
                 sleep(Duration::from_millis(1));
             }
         }
 
+        // Rotate destination keypairs so that the next round of transactions will have different
+        // transaction signatures even when blockhash is reused.
+        dest_keypair_chunks[chunk_index].rotate_left(1);
```

Comment:
Can we just do this with indexing so there is no copying.

Reply:
I see it's copying refs, so not that terrible.

Reply:
Yeah, haven't seen any performance hit from this. The `VecDeque` makes the implementation a lot cleaner